File: ports/csharp/src/core/kernel/BcsEventLoop.cs
using System;

namespace Bcs.Core.Kernel
{
    public class BcsEventLoop : BcsObject
    {
        public BcsEventDispatcher EventDispatcher { get; }
        private readonly object _lock = new object();
        private bool _isRunning = false;
        private int _quitCode = 0;

        public BcsEventLoop(BcsEventDispatcher dispatcher, BcsObject parent = null) : base(parent)
        {
            EventDispatcher = dispatcher ?? new BcsEventDispatcher();
        }

        public int Exec()
        {
            lock (_lock)
            {
                if (_isRunning) return -1;
                _isRunning = true;
            }

            EventDispatcher.Exec();

            lock (_lock)
            {
                _isRunning = false;
                return _quitCode;
            }
        }

        public void Exit(int returnCode = 0)
        {
            lock (_lock)
            {
                _quitCode = returnCode;
            }
            EventDispatcher.Stop();
        }

        public void Quit()
        {
            Exit(0);
        }

        public bool IsRunning
        {
            get { lock (_lock) { return _isRunning; } }
        }
    }
}

File: ports/rust/src/core/kernel/bcs_event_loop.rs
use std::sync::{Arc, Mutex};
use super::bcs_event_dispatcher::BcsEventDispatcher;
use super::bcs_object::BcsObject;

/// BcsEventLoop models QEventLoop
pub struct BcsEventLoop {
    pub base: Arc<BcsObject>,
    pub dispatcher: Arc<BcsEventDispatcher>,
    is_running: Mutex<bool>,
    quit_code: Mutex<i32>,
}

impl BcsEventLoop {
    pub fn new(dispatcher: Arc<BcsEventDispatcher>, parent: Option<Arc<BcsObject>>) -> Self {
        let base = if let Some(p) = parent {
            BcsObject::new_with_parent(p)
        } else {
            BcsObject::new()
        };

        Self {
            base,
            dispatcher,
            is_running: Mutex::new(false),
            quit_code: Mutex::new(0),
        }
    }

    pub fn exec(&self) -> i32 {
        {
            let mut running = self.is_running.lock().unwrap();
            if *running {
                return -1;
            }
            *running = true;
        }

        self.dispatcher.exec();

        let mut running = self.is_running.lock().unwrap();
        *running = false;

        *self.quit_code.lock().unwrap()
    }

    pub fn exit(&self, return_code: i32) {
        *self.quit_code.lock().unwrap() = return_code;
        self.dispatcher.stop();
    }

    pub fn quit(&self) {
        self.exit(0);
    }

    pub fn is_running(&self) -> bool {
        *self.is_running.lock().unwrap()
    }

    pub fn object(&self) -> Arc<BcsObject> {
        self.base.clone()
    }
}
